Tracking block padding when doing allocation tracing

c-odrin opened the pull request:
Summary:
Currently CUDACachingAllocator tracing records logical sizes for allocations(i.e. the amount of bytes that are being requested by the application). We would like record however the block size as well to show padding from the allocator.

For example the following can be sources of padding:
1) Rounding
https://github.com/pytorch/pytorch/blob/6796979ee1063890fd04bbf21f298f669129df8f/c10/cuda/CUDACachingAllocator.cpp#LL1293-L1304C4

2) As well as further padding when doing new allocations(rounding to kSmallBuffer and kRoundLarge):
https://github.com/pytorch/pytorch/blob/6796979ee1063890fd04bbf21f298f669129df8f/c10/cuda/CUDACachingAllocator.cpp#LL1538-L1546C3

3) When blocks are not being split:
https://github.com/pytorch/pytorch/blob/6796979ee1063890fd04bbf21f298f669129df8f/c10/cuda/CUDACachingAllocator.cpp#LL1528-L1536C4

Test Plan: Added test in test_cuda.py

Differential Revision: D41374700

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -878,6 +879,7 @@ class DeviceCachingAllocator {
             TraceEntry::OOM,
             device_free,
             params.size(),
+            0, // not used for TraceEntry:OOM
```
We have both the rounded and unrounded number here, so we should replace params.size() with the unrounded number and used rounded.size() for the rounded number. Previously we only reported the post-rounded number.

c-odrin replied:
Similar to `TraceEntry::ALLOC`, I'm now logging the unrounded size and also storing the `rounded_by` alongside it, to make it consistent and also align with the meaning of the newly added field.

zdevito commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -151,18 +151,23 @@ struct TraceEntry {
         stream_(stream),
-        size_(size) {}
+        size_(size),
+        block_size_(block_size) {}
```
Instead of storing block_size, let's store the amount of amount of memory rounded away `rounded_by = (block_size_ - size_)`. This is more general (it can be adapted to any allocator where an action introduces waste), and it gives meaning to some events like SEGEMENT_ALLOC which don't round off anything at all (and have the value 0). In the future if we encode the events differently to save space this will also give much smaller numbers which will compress better.

c-odrin replied:
Resolved by adding `rounded_by` along `size` in `TraceEntry` struct

zdevito reviewed (commented):
Looks good! A couple minor tweaks to keep size of events down in the future.

zdevito reviewed (approved):
Looks good!

c-odrin commented:
@pytorchbot merge

c-odrin commented:
@pytorchbot merge

c-odrin commented:
@pytorchbot merge

c-odrin commented:
@pytorchbot merge --rebase

c-odrin commented:
@pytorchbot merge --rebase

c-odrin committed:
Tracking block padding when doing allocation tracing (#89429)

Summary:
Currently CUDACachingAllocator tracing records logical sizes for allocations(i.e. the amount of bytes that are being requested by the application). We would like record however the block size as well to show padding from the allocator.

For example the following can be sources of padding:
1) Rounding
https://github.com/pytorch/pytorch/blob/6796979ee1063890fd04bbf21f298f669129df8f/c10/cuda/CUDACachingAllocator.cpp#LL1293-L1304C4

2) As well as further padding when doing new allocations(rounding to kSmallBuffer and kRoundLarge):
https://github.com/pytorch/pytorch/blob/6796979ee1063890fd04bbf21f298f669129df8f/c10/cuda/CUDACachingAllocator.cpp#LL1538-L1546C3

3) When blocks are not being split:
https://github.com/pytorch/pytorch/blob/6796979ee1063890fd04bbf21f298f669129df8f/c10/cuda/CUDACachingAllocator.cpp#LL1528-L1536C4

Pull Request resolved: https://github.com/pytorch/pytorch/pull/89429

Test Plan: Added test in test_cuda.py

Reviewed By: zdevito

Differential Revision: D41374700

Pulled By: c-odrin

fbshipit-source-id: c3c04ad3d99f97f8819217343e71861d27e73982
